hub/core/storage/provider.py

```diff
@@ -0,0 +1,129 @@
+from collections.abc import MutableMapping
+from typing import Optional
+from abc import ABC, abstractmethod
+
+
+class Provider(ABC, MutableMapping):
+    """
+    An abstract base class for implementing a provider.
+    To add a new provider using Provider, create a subclass and implement all 5 abstract methods below.
+    Alternatively, you can inherit from ProviderMapper and have a simpler implementation.
+    """
+
+    @abstractmethod
+    def __getitem__(
+        self,
+        path: str,
+        start_byte: Optional[int] = None,
+        end_byte: Optional[int] = None,
+    ):
+        """
+        Gets the object present at the path.
+        Optionally, if the start_byte and/or end_byte arguments are specified, it only returns required bytes
+
+        Example:
+            local_provider = LocalProvider("/home/ubuntu/Documents/")
+            my_data = local_provider("abc.txt")
+
+        Args:
+            path (str): The path relative to the root of the provider.
+            start_byte (int, optional): If only specific bytes starting from start_byte are required.
+            end_byte (int, optional): If only specific bytes upto end_byte are required.
+
+        Returns:
+            bytes: The bytes of the object present at the path.
+
+        Raises:
+            InvalidBytesRequestedError: If `start_byte` > `end_byte` or `start_byte` < 0 or `end_byte` < 0.
```

Comment:
What's the convention for documentation on ABC's? I feel like this clause can be moved to the children, but I'm not sure if it's standard. Your call

Reply:
I think we can leave them here, would help users who want to implement a subclass of this to understand the expected behavior.